Feat: 이메일, 구독자 페이지 리코일&리액트쿼리 추가

pinomad opened the pull request:
### 주요 구현 사항
1. 이메일 템플릿 목록 화면 적용, 제목 클릭하면 이동하게 변경, 수정 버튼 삭제
2. 이메일 생성 step1 화면 적용
3. 이메일 생성 step2 화면 적용
4. 구독자 설정 화면 적용
5. 구독자 추가 페이지

### 고민거리 및 미구현 사항들
1. 구독자삭제, 이메일삭제 등 삭제 api 요청 후 fetch 하는 로직이 맞는지? +(삭제 컨펌 얼럿창이 필요할거같음)
2. 브라우저 이외의 다른 곳을 클릭했다가 다시 브라우저를 클릭하면 fetch 요청을 왜 다시 하는지 잘 모르겠습니다. 체크박스 테스트시 정상적으로 데이터가 넘어갈때도 있고 아닐때도 있습니다. 아직 제대로 이해를 못한탓 같습니다.
3. 이메일 생성 step1 들어왔을때 내가 선택한 구독자가 그대로 보여야할것같습니다, 아니면 변경사항 없으면 저장 안하던가.. 지금은 선택 안하고 넘어갈때마다 recipients 없는걸로 저장됩니다.
4. step2 에서 이메일 타이틀 연동되게 해야합니다. -> 어떻게 해야하는지 고민입니다. 아톰으로 빼서 관리할지? onChange 할때마다 반영되게 해야하는지?
6. 전체적으로 에러 핸들링 안되어있습니다
7. 이메일 주소는 도메인 포함 30자까지 받으면 될 것 같습니다. 현재 구조는 텍스트가 길면 한줄 더 차지하지 않고 두번쨰 컬럼을 침범하는 형태입니다. 높이값은 변하지 않아서 따로 설정 안해도 될 것 같습니다.
### 추가 커밋 사항
1. 구독자 추가 api 추가했습니다.
2. 비동기부분 async await 처리하였습니다.
3. 수정 후 조회가 아닌 useQuery를 사용하여 상태 업데이트를 적용했습니다.
4. email nav의 emailTitle을 atom으로 추가하여 전역상태로 뺐고 input과 onChange 동기화 했습니다.
5. 이메일 생성 step1 들어왔을때 내가 선택한 구독자가 그대로 보이는 로직은 실패했습니다 시간을 너무 많이쓰는것같아서 일단 나중에 시도하려합니다.

Commit:
feat: 이메일, 구독자 페이지 리코일&리액트쿼리 적용
1. 이메일 템플릿 목록 화면 적용, 제목 클릭하면 이동하게 변경, 수정 버튼 삭제
2. 이메일 생성 step1 화면 적용
3. 이메일 생성 step2 화면 적용
4. 구독자 설정 화면 적용

Commit:
chore: 버전업

roadzmoon76 commented on src/pages/emails/EmailEditingStep02.jsx:
```diff
@@ -1,32 +1,86 @@
+    },
+    onSuccess: emailTemplateInfoData => {
+      const { emailTitle, sender, emailPreviewText } = emailTemplateInfoData;
+      setSendingInfo({
```
띄워쓰기~

roadzmoon76 commented on src/pages/emails/EmailEditingStep01.jsx:
```diff
@@ -106,15 +74,15 @@ export default function EmailEditingStep01() {
     const removedIsCheckedSubscribers = removeIsCheckedProperty(checkedRows);
+    const emailTemplateData = {
+      editingStep: '02',
+      recipients: removedIsCheckedSubscribers,
```
위의 코드들에선 `recipients: removeIsCheckedProperty(checkedRows)` 이런 느낌으로 쓰셨던 것 같은데 갑자기 이렇게 쓰신 이유가 있나요?

pinomad replied:
결국 이 화면에서 보내주는값은 최종적으로 이메일 템플릿 데이터이고 editingStep도 포함되어있어서 저렇게 네이밍했습니다

roadzmoon76 replied:
아뇨 변수명을 말씀드린게 아니라 위의 코드들에서 setState 등을 보면 유틸함수 실행을 인자안에서 바로 하는 식으로 쓰셨다가 갑자기 밑에선 또 유틸함수를 적용한 값을 따로 변수에 담아두는 식으로 써서 컨벤션이 일정해보이지 않는다는 것을 말씀드린거에요.

pinomad replied:
이 부분은 이해가 잘 안가서 허들로 이야기 하면 좋을것 같습니다.

roadzmoon76 commented on src/pages/emails/EmailEditingStep01.jsx:
```diff
@@ -106,15 +74,15 @@ export default function EmailEditingStep01() {
-    console.log('이메일 수정 api 실행');
-
-    const emailId = 'test';
+    fetchUpdateEmail(userId, param.email_id, emailTemplateData);
```
handleNextClick 함수를  async 로 만들고 await 를 붙여줘야 fetch 가 의도대로 동작할 듯 합니다.

roadzmoon76 commented on src/pages/emails/EmailEditingStep02.jsx:
```diff
@@ -1,32 +1,86 @@
+    return <Error>error</Error>;
+  }
 
   const handlePrevClick = () => {
```
async (위의 코멘트와 중복 되는 부분은 짧게 키워드만 쓸게요)

roadzmoon76 commented on src/pages/emails/EmailEditingStep02.jsx:
```diff
@@ -1,32 +1,86 @@
+    };
 
-    const emailId = 'test';
+    fetchUpdateEmail(userId, param.email_id, emailTemplateData);
```
await

roadzmoon76 commented on src/pages/emails/EmailEditingStep02.jsx:
```diff
@@ -1,32 +1,86 @@
-  const [inputPreviewText, setInputPreviewText] = useState('');
+  const param = useParams();
+  const userId = useRecoilValue(userIdAtom);
+  const [sendingInfo, setSendingInfo] = useState({
```
emailTitle 은 navbar 의 타이틀과 동기화가 되도록 recoil 을 활용하셔야 할 것 같습니다.

다른 step 에서도 처음 데이터를 fetching 해왔을 때 해당 아톰을 최신의 제목으로 업데이트 해주는 로직도 필요하구요.

pinomad replied:
반영하였습니다

roadzmoon76 replied:
step01 에서 이메일 정보 불러왔을 때 해당 이메일 제목이 nav 제목에 반영되는 로직이 안보이는 것 같습니다.

roadzmoon76 commented on src/pages/emails/Emails.jsx:
```diff
@@ -52,15 +51,19 @@ export default function Emails() {
+  const handleDeleteButtonClick = async emailId => {
+    await fetchDeleteEmail(userId, emailId);
+
+    const deletedEmailTemplates = await fetchEmails(userId);
```
변수명이 적절해보이지 않습니다. 새로 업데이트 된 이메일 리스트의 느낌을 주는 변수명이 좋을 것 같아요.
`newEmailTemplates` 정도도 좋을 것 같습니다

또 새롭게 정보를 받아오는 로직은 쿼리를 그대로 활용하는게 좋을 것 같습니다.

저는 기본값이 0인 count 라는 스테이트를 만들고, 해당 count 스테이트를 querykey 에 dependency 로 넣어줬고 삭제가 완료되면 `setCount(count + 1)` 이라는 걸 수행하게 해서 삭제된 것을 감지하고 쿼리가 새로운 데이터를 받아올 수 있도록 했습니다.

pinomad replied:
반영하였습니다

roadzmoon76 commented on src/pages/emails/Emails.jsx:
```diff
@@ -52,15 +51,19 @@ export default function Emails() {
 
-  const handleEditButtonClick = () => {
-    console.log('수정 버튼 클릭 -> 해당 step으로 가야함');
+  const handleEditButtonClick = (emailId, editingStep) => {
```
수정 버튼은 삭제했으니 핸들러 이름도 바꾸셔야 할듯 합니다.

roadzmoon76 commented on src/pages/subscribers/Subscribers.jsx:
```diff
@@ -1,93 +1,60 @@
+    return <Error>error</Error>;
+  }
+
   const handleDeleteSubscriberButtonClick = () => {
```
async

roadzmoon76 commented on src/pages/subscribers/Subscribers.jsx:
```diff
@@ -1,93 +1,60 @@
-    console.log(removedIsCheckedSubscribers); // 구독자 삭제 api 요청, delete하고 리턴으로 조회한 값 받아옴
-
-    // setAddedIsCheckedSubscribers() 구독자 삭제후 setState()
+    fetchDeleteSubscribers(userId, subscribersData);
```
await

roadzmoon76 replied:
구독자 삭제후 위에 제가 커멘트 남겨드린 방법을 이용해서 새로운 구독자 정보를 받아올 수 있도록 해야할 것 같습니다.

pinomad replied:
반영하였습니다

roadzmoon76 reviewed (changes requested):
## 정진님의 고민거리들에 대한 답변
1. 삭제 요청을 한뒤 페이지의 정보를 받아오는 query 를 이용해 다시 새로운 정보를 받아오도록 하게 하면 됩니다. 코멘트에 써놨으니 참고해주세요
2. `useQuery` 의 `refetchOnWindowFocus` 기본값이 true 이기 때문입니다. 체크 박스 오류는 async, await 을 안써서 fetch 가 제대로 반영이 안되서 그런 것 같습니다.
3. 해당 부분은 사용자의 구독자 목록과 이메일의 수신자 목록을 둘다 불러와서 구현해야겠네요
4. 아톰으로 관리하고 onChange 마다 동기화 되게 해야합니다. 또 다른 step 에선 이메일 정보 불러올 때 해당 아톰도 최신 이메일 제목을 반영하도록 해야합니다.
5. 에러핸들링은 해야 할곳들 정진님이 따로 메모해두시고 나중에 한번에 같이 하죠. 애초에 서버에서 보내주는 에러도 제대로 다듬어져있지 않아서 나중에 한번에 손봐야 할듯 합니다.
6. 두번째 컬럼을 침범한다는게 텍스트가 겹친다는건가요? 어떤 이유에서든 기존 레이아웃이 무너지면 안됩니다. 테스트가 길어져서 한줄 더 차지하더라도 높이가 늘어나지 않게 하는게 맞는 것 같습니다.

Commit:
fix: 비동기부분 async-await 적용 및 emailTile atom 추가
1. 구독자 추가 api 추가했습니다.
2. 비동기부분 async await 처리하였습니다.
3. 수정 후 조회가 아닌 useQuery를 사용하여 업데이트를 적용했습니다.
4. email nav의 emailTitle을 atom으로 추가했습니다.

roadzmoon76 committed:
feat: 이메일 편집 step03 정적인 페이지 추가

1. 이메일 편집 step03 정적인 페이지를 추가했습니다.

pinomad reviewed (commented):
1. 구독자 추가 api 추가했습니다.
2. 비동기부분 async await 처리하였습니다.
3. 수정 후 조회가 아닌 useQuery를 사용하여 업데이트를 적용했습니다.
4. email nav의 emailTitle을 atom으로 추가했습니다.
5. 이메일 생성 step1 들어왔을때 내가 선택한 구독자가 그대로 보이는 로직은 실패했습니다 시간을 너무 많이쓰는것같아서 일단 나중에 시도하려합니다.

roadzmoon76 commented on src/recoil/emailTtile/atom.js:
```diff
@@ -0,0 +1,8 @@
+import { atom } from 'recoil';
```
폴더명 실수하신 것 같습니다.

pinomad replied:
앗 넵 수정했습니다

roadzmoon76 reviewed (changes requested):
파일명 실수하신 것 수정 부탁드립니다.

또 코멘트에도 남겼지만 step01 에서 이메일 불러왔을 때 해당 이메일 제목으로 nav 제목 업데이트하는 로직이 안보입니다.

구독자 일치 로직은 정진님 말씀대로 팀 메모에 적어두고 나중에 반영하죠

Commit:
fix: 폴더명 수정 및 nav바 타이틀 동기화

pinomad committed:
Merge branch 'develop' into feature/add-recoil&reactquery-email-page

pinomad reviewed (commented):
이메일 step1단계 말씀하신부분 반영하였습니다.

roadzmoon76 reviewed (approved):
수고하셨습니다

Commit:
fix: 유틸함수 컨벤션 수정
